lib/src/core/mod.rs

```diff
@@ -0,0 +1,56 @@
+use anyhow::{anyhow, Context, Result};
+extern crate serde;
+use self::serde::{Deserialize, Serialize};
+use crate::*;
+use async_trait::async_trait;
+
+mod proxy;
+pub mod verifier;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct Evidence {
+    pub tee: String,
+    pub quote: String,
+    pub ehd: String,
```

Comment:
The KBS needs access the nonce field of the evidence to make sure that the correct nonce was returned from the TEE. In the RFC the nonce is part of the `ehd` but here we don't explicitly specify the format of the `ehd`. We might want to create an `ehd` struct that defines the `ehd` field of the evidence struct so that we guarantee the nonce is always in the same place. Another option would be to add a `nonce` method to `TEETraits` that the KBS could call to get the nonce.

Reply:
OK. I will extend the `sample` TEE codes to add the `ehd` definition example.
We can ensure the SGX, TDX, SEV-SNP attestation needs the nonce, but can't sure whether any new TEE will also depends on the nonce, so I think it's inapporiate to add the `ehd` explicit definition in this `core/mod.rs`.

Reply:
If some future architecture comes along, it must support a nonce or it won't work with the KBS. If there is no nonce, there is no way for the KBS to match the challenge to the evidence. We need to be flexible about the format of the evidence, but I think it is fine and maybe even good if we require all future evidence to include a nonce.

Reply:
I think i can catch your point, so I tried to update the Evidence struct as following:
```
// core/mod.rs

#[derive(Serialize, Deserialize, Debug)]
pub struct Ehd {
    // The mandate nonce
    pub nonce: String,
    // The TEE specific enclave hold data
    pub ehd: String,
}
#[derive(Serialize, Deserialize, Debug)]
pub struct Evidence {
    pub tee: String,
    pub quote: String,
    pub aad: String,
    pub ehd: Ehd,
}
```
But this will cause the error: _Deserialize Evidence error, expected struct Ehd at line 1 column 116)_ during the input evidence's deserialize.
This means we need to parse the json evidence manully to construct the `Evidence` struct. So I think we need to keep the `Ehd` struct's definition in each specific verifier.

Reply:
I was thinking about this more and we might actually need to do something similar for the TEE public key. The KBS needs to access to the key in order to encrypt the secrets, but we also have to make sure that the key that the KBS thinks is the public key of the TEE matches the public key in the evidence. Every platform must have some way to link the TEE public key to the attestation report (just like the nonce).

Interestingly the nonce and public key are both provided in two places. 

First, the KBS protocol specifies the attestation as 
```
{
    "evidence": {
	    "nonce": "",
	    "tee-pubkey": {
		    "algorithm": "",
		    "pubkey-length": "",
		    "pubkey": ""
		},
	    "tee-evidence": {}
	}
}
```
Then, the AS gives this example of `evidence`
```
{
    "tee" : "tdx",
    "quote": <Base64 encoded quote>,
    // The ehd's hash will be embeded in Quote->report_data.
    "ehd": 
    {
        "nonce" : "cdfunsas23445xd",
        "public-key" : "xxxxxxxx"
    },
    "aad": 
    {
        "tdelInfo" : <Base64 encoded eventlog's information>,
        "tdelData" : <Base64 encoded eventlog's data>
    }
}
```

In order for this to be secure, I think that the KBS or the AS has to make sure that the nonce and public-key in the evidence exactly match the nonce and the public key in the attestation. There are two options. First, the KBS could check that the values match. For this to work we need to make sure that the evidence always includes the nonce and the public key in the same place. This is what I first suggested. There is another options that might be nicer. We could have the AS check that the values match instead of the KBS. All we would need to do is pass the entire attestation payload into the AS rather than just the evidence. if we pass the entire attestation payload (which btw should probably include an `additional-params` field) into the AS, we don't have to mandate the format of the evidence at all. We just have to make sure that the AS checks that the nonce and public key match the (signed) nonce and public key in the attestation report. 

Also note that in the example above the evidence does not contain the same key metadata that the attestation does. Is this an issue?

Reply:
Yes, I think it's good idea to pass the entire attestation payload to the AS. And the `additional-params`  can be put into the TEE specific `tee-evicence` JSON string. And jiale has added a new "tee" to designate the Evidence's type, as:
```JSON
{
    "evidence": {
        "nonce": "",
         /* TEE type name */
	"tee": "",
	"tee-pubkey": {
            "algorithm": "",
            "pubkey-length": "",
            "pubkey": ""
        },
	"tee-evidence": {}
    }
}
```